### PR TITLE
iotkit: initial commit of the hello world

### PR DESCRIPTION
**Pull request by rakyll**

### CODE REVIEW

**Comment by rakyll** [comment]
/cc @broady 

I will also follow with a tutorial how to wire things, build the program and deploy it on the device.

**Comment by broady** [comment]
Is this related to GCP? Doesn't look like it?

**Comment by rakyll** [comment]
This is going to be a sample for [GCP IoT Kit](https://cloud.google.com/solutions/iot/kit/). The scope is more like "how to build" for the board for now. We will extend the samples with GCP service integrations in the future.

File: iotkit/helloworld/display/display.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+	line2 = 0x08
+	line1 = 0x00
+
+	dots10 = 0x04
+	dots8  = 0x00
+
+	// addresses
+	lcdAddr = 0x3e
+	rgbAddr = 0x62
```
Comment (broady) [inline iotkit/helloworld/display/display.go]: Are all/some of these constants something that should live in an `cloud/iokit` package?

File: iotkit/helloworld/display/display.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+		return nil, fmt.Errorf("cannot open LCD device: %v", err)
+	}
+	rgb, err := i2c.Open(o, rgbAddr)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open RGB device: %v", err)
+	}
+
+	time.Sleep(50 * time.Millisecond) // wake up time after opening
+
+	// LCD function set, try a few times
```
Comment (broady) [inline iotkit/helloworld/display/display.go]: Why? More comments explaining what's going on here would be good.

File: iotkit/helloworld/display/display.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+// SetText clears the screen and prints the given text on the display.
+func (d *Device) SetText(text string) error {
+	if err := d.Clear(); err != nil {
+		return err
+	}
+	time.Sleep(10 * time.Millisecond)
+
+	// num lines = 2
+	if err := d.lcd.Write([]byte{0x80, 0x20 | 0x08}); err != nil {
```
Comment (broady) [inline iotkit/helloworld/display/display.go]: What are these bytes/flags?

File: iotkit/helloworld/display/display.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+		return err
+	}
+	time.Sleep(10 * time.Millisecond)
```
Comment (broady) [inline iotkit/helloworld/display/display.go]: Why 10 ms?

File: iotkit/helloworld/display/display.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	// return home
+	if err := d.lcd.Write([]byte{0x80, returnHome}); err != nil {
+		return err
+	}
+	time.Sleep(10 * time.Millisecond)
+
+	var row int
+	var col int
+	for _, c := range text {
+		if c == '\n' || col == 16 {
```
Comment (broady) [inline iotkit/helloworld/display/display.go]: Write a comment. This is wrapping text, right?

File: iotkit/helloworld/display/display.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+func (d *Device) Clear() error {
+	return d.lcd.Write([]byte{0x80, 0x01}) // clear display
+}
+
+// SetRGB sets the backlight to the given color.
+func (d *Device) SetRGB(r, g, b int) error {
+	cmds := [][]byte{
+		{0, 0},
+		{1, 0},
+		{0x08, 0xaa},
```
Comment (broady) [inline iotkit/helloworld/display/display.go]: Reference for this sequence would be good, too.

File: iotkit/helloworld/display/display.go (hunk 1 above)
Comment (rakyll) [inline iotkit/helloworld/display/display.go]: No, these are specific to this device and never changes.

**Comment by rakyll** [comment]
The links to the actual datasheets are broken, see the final page at http://www.mouser.com/catalog/specsheets/Seeed_104030001.pdf. I will use their reference Python code to modify and document the initialization, see https://github.com/DexterInd/GrovePi/tree/master/Software/Python/grove_rgb_lcd.

File: iotkit/helloworld/display/display.go (hunk 6 above)
Comment (rakyll) [inline iotkit/helloworld/display/display.go]: Added a link to the existing Python reference. That's so far the only the good reference for this specific device.

**Pull request merged**